doc/mgmt/SONiC_YANG_Model_Guidelines.md

```diff
@@ -0,0 +1,961 @@
+.....
+.....
+```
+#### YANG
+```
+.....
+container ACL_TABLE {
+	list ACL_TABLE_LIST {
+		key aclname;
+			
+		leaf aclname {
+			type string;
+		}
+		.....
+		.....
+	}
+}
+container ACL_RULE {
+	list ACL_RULE_LIST {
+		key "aclname rulename";
+		
+		leaf aclname {
+			type leafref {
+				path "../../../ACL_TABLE/ACL_TABLE_LIST/aclname"; //Refers to other table 'ACL_TABLE'
+			}
+		}
+		
+		leaf rulename {
+			type string;
+		}
+		.....
+		.....
+	}
+}
+```
+
+### 12. Mapping tables in Redis are defined using nested 'list'. Use 'sonic-ext:map-list "true";' to indicate that the 'list' is used for mapping table. The outer 'list' is used for multiple instances of mapping. The inner 'list' is used for mapping entries for each outer list instance.
+
+Example :
+
+#### ABNF
+```
+; TC to queue map
+;SAI mapping - qos_map with SAI_QOS_MAP_ATTR_TYPE == SAI_QOS_MAP_TC_TO_QUEUE. See saiqosmaps.h
+key                    = "TC_TO_QUEUE_MAP_TABLE:"name
+;field
+tc_num = 1*DIGIT
+;values
+queue  = 1*DIGIT; queue index
+
+```
+
+#### YANG
+```
+	container TC_TO_QUEUE_MAP {
+		list TC_TO_QUEUE_MAP_LIST {
+			key "name";
+			sonic-ext:map-list "true"; //special annotation for map table
+
+			leaf name {
+				type string;
+			}
+
+			list TC_TO_QUEUE_MAP_LIST { //this is list inside list for storing mapping between two fields
+				key "tc_num qindex";
+				leaf tc_num {
+					type string {
+						pattern "[0-9]?";
+					}
+				}
+				leaf qindex {
+					type string {
+						pattern "[0-9]?";
+					}
+				}
+			}
+		}
+	}
+```
+
+
+### 13. 'ref_hash_key_reference' in ABNF schema is defined using 'leafref' to the referred table.
+
+Example : 
+
+#### ABNF
+```
+; QUEUE table. Defines port queue.
+; SAI mapping - port queue.
+
+key             = "QUEUE_TABLE:"port_name":queue_index
+queue_index     = 1*DIGIT
+port_name       = ifName
+queue_reference = ref_hash_key_reference
+
+;field            value
+scheduler    = ref_hash_key_reference; reference to scheduler key
+wred_profile = ref_hash_key_reference; reference to wred profile key
+
+```
+
+#### YANG
+```
+container sonic-queue {
+	container QUEUE {
+		list QUEUE_LIST {
+			key "ifname qindex";
+
+			leaf ifname {
+				type leafref {
+					path "/prt:sonic-port/prt:PORT/prt:ifname";
+				}
+			}
+
+			leaf qindex {
+				type string {
+					pattern "[0-8]((-)[0-8])?";
+				}
+			}
+
+			leaf scheduler {
+				type leafref {
+					path "/sch:sonic-scheduler/sch:SCHEDULER/sch:name"; //Reference to SCHEDULER table
+				}
+			}
+
+			leaf wred_profile {
+				type leafref {
+					path "/wrd:sonic-wred-profile/wrd:WRED_PROFILE/wrd:name"; // Reference to WRED_PROFILE table
+				}
+			}
+		}
+	}
+}
+```
+
+### 14. To establish complex relationship and constraints among multiple tables use 'must' expression. Define appropriate error message for reporting to Northbound when condition is not met. For existing feature, code logic could be  reference point for deriving 'must' expression.
```

Comment:
Generic.

Reply:
Though 'must' expression is not mandatory for ABNF to YANG modelling, it is needed for semantic validation, intention is to make sure that developer understands it.